ddtrace/internal/sampling.py

```diff
@@ -178,6 +193,66 @@ def apply_span_sampling_tags(self, span):
             span.set_metric(_SINGLE_SPAN_SAMPLING_MAX_PER_SEC, self._max_per_second)
 
 
+def get_span_sampling_rules():
+    # type: () -> List[SpanSamplingRule]
+    json_rules_raw = os.getenv("DD_SPAN_SAMPLING_RULES")
+    if json_rules_raw is None:
+        return []
+    else:
+        sampling_rules = []
+        try:
+            json_rules = json.loads(json_rules_raw)  # type: List[SpanSamplingRules]
+            if not isinstance(json_rules, list):
+                raise TypeError("DD_SPAN_SAMPLING_RULES is not list, got %r" % json_rules)
+        except JSONDecodeError:
+            raise ValueError("Unable to parse DD_SPAN_SAMPLING_RULES=%r" % json_rules_raw)
+        for rule in json_rules:
+            if not isinstance(rule, dict):
+                raise TypeError("rule specified via DD_SPAN_SAMPLING_RULES is not a dictionary:%r" % rule)
+            # If sample_rate not specified default to 100%
+            sample_rate = float(rule.get("sample_rate", 1.0))
+            service = rule.get("service")
+            name = rule.get("name")
+            # If max_per_second not specified default to no limit
+            max_per_second = int(rule.get("max_per_second", -1))
+            if service is None and name is None:
+                raise ValueError("Neither service or name specified for single span sampling rule:%r" % rule)
+            if service:
+                _check_unsupported_pattern(service)
+            if name:
+                _check_unsupported_pattern(name)
+
+            try:
+                sampling_rule = SpanSamplingRule(
+                    sample_rate=sample_rate, service=service, name=name, max_per_second=max_per_second
+                )
+            except Exception as e:
+                raise ValueError("Error creating single span sampling rule {}: {}".format(json.dumps(rule), e))
+            sampling_rules.append(sampling_rule)
+        return sampling_rules
+
+
+def _check_unsupported_pattern(string):
+    # type: (str) -> None
+    # We don't support pattern bracket expansion or escape character
+    unsupported_chars = {"[", "]", "\\"}
+    for char in string:
+        if char in unsupported_chars:
+            raise ValueError("Unsupported Glob pattern found, character:%r is not supported" % char)
+
+
+if sys.version_info >= (3, 7, 0):
+    SpanSamplingRules = TypedDict(
+        "SpanSamplingRules",
+        {
+            "name": NotRequired[str],
+            "service": NotRequired[str],
+            "sample_rate": NotRequired[float],
+            "max_per_second": NotRequired[int],
+        },
+    )
```

Comment:
```suggestion
        "SpanSamplingRules",
        {
            "name": str,
            "service": str,
            "sample_rate": float,
            "max_per_second": int,
        },
        total=False,
    )
```

Since all the fields are not required we can simply set total to false: https://peps.python.org/pep-0589/#totality. With this change we can avoid using `NotRequired`

Reply:
[NotRequired](https://peps.python.org/pep-0655/#abstract) is useful when some keys are required and others are not

Reply:
I think I'd prefer `NotRequired` here since it's more explicit. If a new entry is added to the dictionary then it won't implicitly be `NotRequired`. Just my opinion though.

Reply:
Given Brett's comment here: https://github.com/DataDog/dd-trace-py/pull/3977#discussion_r936585840 I think it'd be better to go with using Munir's suggestion above.